Compfiles/Imo1970P4.lean

```diff
@@ -21,15 +20,781 @@ equals the product of the nubmers in the other set.
+    by_contra Hhh
+    simp_all only [Nat.two_dvd_ne_zero, not_or]
+    omega
+
+lemma exactly_three_odd_numbers (n : ℕ) (odd_s : Finset ℕ)
+                                (odd_s_eq: odd_s = (Finset.Icc n (n + 5)).filter (λ x => Odd x)): (odd_s).card = 3 := by
+  -- ∃ (a b c : ℕ), {a, b, c} = odd_s ∧ odd_s.card = 3
+  obtain ⟨x, y, z, ⟨left, ⟨y_eq, z_eq⟩⟩⟩ := odd_props n odd_s odd_s_eq
+  have := (@Finset.card_eq_three ℕ odd_s).mpr
+  apply this
+  use x
+  use x + 2
+  use x + 2 + 2
+  constructor
+  · omega
+  · constructor
+    · omega
+    · constructor
+      · omega
+      · simp_all only
+
+
+
+
+
+
+lemma at_most_one (n : ℕ) (x y : ℕ)
+  (x_in_interval : x ∈ Finset.Icc n (n + 5)) (y_in_interval : y ∈ Finset.Icc n (n + 5))
+  (x_div_5 : 5 ∣ x )
+  (x_non_div_2 : ¬ 2 ∣ x)
+  (y_div_5 : 5 ∣ y )
+  (y_non_div_2 : ¬ 2 ∣ y) :
+  x = y := by
+  simp_all only [Finset.mem_Icc, Nat.two_dvd_ne_zero]
+  omega
+
+lemma card_of_equal (s : Finset ℕ) : (∀ x ∈ s, ∀ y ∈ s,  x = y) → s.card ≤ 1 := by
+  exact (@Finset.card_le_one ℕ s).mpr
+
+lemma five_divides_odd_at_most_once (n : ℕ) (s odd_s : Finset ℕ) (partition : s = Finset.Icc n (n + 5))
+                                    (odd_s_eq: odd_s = s.filter (λ x => Odd x)) : (odd_s.filter (λ x => 5 ∣ x)).card ≤ 1 := by
+  have : (∀ x ∈ (odd_s.filter (λ x => 5 ∣ x)), ∀ y ∈ (odd_s.filter (λ y => 5 ∣ y)), x = y) → (odd_s.filter (λ x => 5 ∣ x)).card ≤ 1 := by
+    exact (card_of_equal) (odd_s.filter (λ x => 5 ∣ x))
+
+  apply this
+
+  rw[odd_s_eq]
+
+  intro x x_in
+  intro y y_in
+
+  simp only [Finset.mem_filter] at x_in
+  obtain ⟨⟨X1, X2⟩, X3⟩ := x_in
+
+  simp only [Finset.mem_filter] at y_in
+  obtain ⟨⟨Y1, Y2⟩, Y3⟩ := y_in
+
+  apply at_most_one n
+  rw[partition] at X1
+  exact X1
+  rw[partition] at Y1
+  exact Y1
+  exact X3
+  exact Odd.not_two_dvd_nat X2
+  exact Y3
+  exact Odd.not_two_dvd_nat Y2
+
+lemma unique_divisor (n : ZMod 3) (a b c : ℕ) (n_eq_a : n = a) (s : Finset ℕ) (s_eq : s = ({a, b, c} : Finset ℕ)) (b_eq: b = a + 2) (c_eq : c = b + 2) : ∃! a', a' ∈ s ∧ 3 ∣ a' := by
+  fin_cases n
+  · use a
+    have three_div_a : 3 ∣ a := by
+      apply (ZMod.natCast_zmod_eq_zero_iff_dvd a 3).mp
+      simp_all only [Fin.zero_eta]
+    constructor
+    · simp
+      constructor
+      · aesop
+      · simp_all only [Fin.zero_eta]
+    · intro o
+      rintro ⟨o_in_s, three_div_o⟩
+      rw[s_eq] at o_in_s
+      simp_all only [Fin.zero_eta, Finset.mem_insert, Finset.mem_singleton]
+      omega
+  · use b
+    have three_div_b : 3 ∣ b := by
+      simp_all only [Fin.mk_one]
+      apply (ZMod.natCast_zmod_eq_zero_iff_dvd (a + 2) 3).mp
+      simp_all only [Nat.cast_add, Nat.cast_ofNat]
+      rw[← n_eq_a]
+      reduce_mod_char
+    constructor
+    · simp_all only [Fin.mk_one, Finset.mem_insert, add_right_eq_self, OfNat.ofNat_ne_zero, Finset.mem_singleton, self_eq_add_right, or_false, or_true, and_self]
+    · intro o
+      rintro ⟨o_in_s, three_div_o⟩
+      rw[s_eq] at o_in_s
+      simp_all
```

Comment:
Nonterminal simp should have an `only`.

Reply:
I've resolved this and all the others (I actually used a wrong regex to search for those earlier).